nomad/host_volume_endpoint.go

```diff
@@ -0,0 +1,515 @@
+		if err != nil {
+			mErr = multierror.Append(mErr, err)
+		} else {
+			raftArgs.Volumes = append(raftArgs.Volumes, vol)
+		}
+	}
+
+	// if we created or updated any volumes, apply them to raft.
+	var index uint64
+	if len(raftArgs.Volumes) > 0 {
+		_, index, err = v.srv.raftApply(structs.HostVolumeRegisterRequestType, raftArgs)
+		if err != nil {
+			v.logger.Error("raft apply failed", "error", err, "method", "register")
+			mErr = multierror.Append(mErr, err)
+		}
+	}
+
+	reply.Volumes = raftArgs.Volumes
+	reply.Index = index
+	return helper.FlattenMultierror(mErr)
+}
+
+func (v *HostVolume) Register(args *structs.HostVolumeRegisterRequest, reply *structs.HostVolumeRegisterResponse) error {
+
+	authErr := v.srv.Authenticate(v.ctx, args)
+	if done, err := v.srv.forward("HostVolume.Register", args, args, reply); done {
+		return err
+	}
+	v.srv.MeasureRPCRate("host_volume", structs.RateMetricWrite, args)
+	if authErr != nil {
+		return structs.ErrPermissionDenied
+	}
+	defer metrics.MeasureSince([]string{"nomad", "host_volume", "register"}, time.Now())
+
+	allowVolume := acl.NamespaceValidator(acl.NamespaceCapabilityHostVolumeRegister)
+	aclObj, err := v.srv.ResolveACL(args)
+	if err != nil {
+		return err
+	}
+
+	if len(args.Volumes) == 0 {
+		return fmt.Errorf("missing volume definition")
+	}
+
+	for _, vol := range args.Volumes {
+		if vol.Namespace == "" {
+			vol.Namespace = args.RequestNamespace()
+		}
+		if !allowVolume(aclObj, vol.Namespace) {
+			return structs.ErrPermissionDenied
+		}
+	}
+
+	// ensure we only try to create valid volumes or make valid updates to
+	// volumes
+	validVols, err := v.validateVolumeUpdates(args.Volumes)
+	if err != nil {
+		return err
+	}
+
+	raftArgs := &structs.HostVolumeRegisterRequest{
+		Volumes:      validVols,
+		WriteRequest: args.WriteRequest,
+	}
+
+	var mErr *multierror.Error
+	var index uint64
+	if len(raftArgs.Volumes) > 0 {
+		_, index, err = v.srv.raftApply(structs.HostVolumeRegisterRequestType, raftArgs)
+		if err != nil {
+			v.logger.Error("raft apply failed", "error", err, "method", "register")
+			mErr = multierror.Append(mErr, err)
+		}
+	}
+
+	reply.Volumes = raftArgs.Volumes
+	reply.Index = index
+	return helper.FlattenMultierror(mErr)
+}
+
+func (v *HostVolume) validateVolumeUpdates(requested []*structs.HostVolume) ([]*structs.HostVolume, error) {
+
+	now := time.Now().UnixNano()
+
+	vols := []*structs.HostVolume{}
+
+	snap, err := v.srv.State().Snapshot()
+	if err != nil {
+		return nil, err
+	}
+
+	var mErr *multierror.Error
+	for _, vol := range requested {
+		vol.ModifyTime = now
+
+		if vol.ID == "" {
+			vol.ID = uuid.Generate()
+			vol.CreateTime = now
+		}
+
+		// if the volume already exists, we'll ensure we're validating the
+		// update
+		current, err := snap.HostVolumeByID(nil, vol.Namespace, vol.ID, false)
+		if err != nil {
+			mErr = multierror.Append(mErr, err)
+			continue
+		}
+		if err = vol.Validate(current); err != nil {
```

Comment:
Yeah, because we'll need to validate the volume even if it's a new one, so having validation for both new and updated volumes together seems nicer. Keeping in mind that `HostVolume.Validate` is no-op for the moment until we get a little further along, of course. :grin:

Reply:
Adding a TODO comment in the `Validate` method for this just to make sure it doesn't fall thru the cracks.